source/extensions/filters/http/peer_metadata/filter.cc

```diff
@@ -0,0 +1,344 @@
+        const auto address = upstream_host->address();
+        switch (address->type()) {
+        case Network::Address::Type::Ip:
+          peer_address = upstream_host->address();
+          break;
+        case Network::Address::Type::EnvoyInternal:
+          if (upstream_host->metadata()) {
+            const auto& filter_metadata = upstream_host->metadata()->filter_metadata();
+            const auto& it = filter_metadata.find("tunnel");
+            if (it != filter_metadata.end()) {
+              const auto& destination_it = it->second.fields().find("destination");
+              if (destination_it != it->second.fields().end()) {
+                peer_address = Network::Utility::parseInternetAddressAndPortNoThrow(
+                    destination_it->second.string_value(), /*v6only=*/false);
+              }
+            }
+          }
+          break;
+        default:
+          break;
+        }
+      }
+    }
+  }
+  const auto metadata_object = metadata_provider_->GetMetadata(peer_address);
+  if (metadata_object) {
+    return Istio::Common::convertWorkloadMetadataToFlatNode(metadata_object.value());
+  }
+  return {};
+}
+
+absl::optional<PeerInfo> MXMethod::derivePeerInfo(const StreamInfo::StreamInfo&,
+                                                  Http::HeaderMap& headers) const {
+  const auto peer_id_header = headers.get(Headers::get().ExchangeMetadataHeaderId);
+  absl::string_view peer_id =
+      peer_id_header.empty() ? "" : peer_id_header[0]->value().getStringView();
+  const auto peer_info_header = headers.get(Headers::get().ExchangeMetadataHeader);
+  absl::string_view peer_info =
+      peer_info_header.empty() ? "" : peer_info_header[0]->value().getStringView();
+  if (!peer_info.empty()) {
+    return lookup(peer_id, peer_info);
+  }
+  return {};
+}
+
+void MXMethod::remove(Http::HeaderMap& headers) const {
+  headers.remove(Headers::get().ExchangeMetadataHeaderId);
+  headers.remove(Headers::get().ExchangeMetadataHeader);
+}
+
+absl::optional<PeerInfo> MXMethod::lookup(absl::string_view id, absl::string_view value) const {
+  // This code is copied from:
+  // https://github.com/istio/proxy/blob/release-1.18/extensions/metadata_exchange/plugin.cc#L116
+  if (max_peer_cache_size_ > 0 && !id.empty()) {
+    auto it = cache_.find(id);
+    if (it != cache_.end()) {
+      return it->second;
+    }
+  }
+  const auto bytes = Base64::decodeWithoutPadding(value);
+  google::protobuf::Struct metadata;
+  if (!metadata.ParseFromString(bytes)) {
+    return {};
+  }
+  const auto fb = ::Wasm::Common::extractNodeFlatBufferFromStruct(metadata);
+  std::string out(reinterpret_cast<const char*>(fb.data()), fb.size());
+  if (max_peer_cache_size_ > 0 && !id.empty()) {
+    // do not let the cache grow beyond max cache size.
+    if (static_cast<uint32_t>(cache_.size()) > max_peer_cache_size_) {
+      cache_.erase(cache_.begin(), std::next(cache_.begin(), max_peer_cache_size_ / 4));
+    }
+    cache_.emplace(id, out);
```

Comment:
Not sure if this matters or not, but if id is already in the cache, this might unnecessarily cause the cache to clear if id already exists in the cache.

Maybe:
```cpp
if(!id.empty()){
  auto id_it = cache_.find(id);
  if (id_it != cache.end()){
      id_it->second = out;
      return;
  }
  if (max_peer_cache_size_ > 0 && !id.empty()) {
    // do not let the cache grow beyond max cache size.
    if (static_cast<uint32_t>(cache_.size()) > max_peer_cache_size_) {
      cache_.erase(cache_.begin(), std::next(cache_.begin(), max_peer_cache_size_ / 4));
    }
    cache_.emplace(id, out);
}
return out;
```

Might be bikeshedding.  Up to you if you want to change it.

Reply:
Same reason as above: backwards compatibility https://github.com/istio/proxy/blob/master/extensions/metadata_exchange/plugin.cc#L133. There's no proper automated test for this caching logic so I am hesitant to touch it in any way.

Reply:
The problem is thread management, will send a follow-up.